kolibri/plugins/setup_wizard/assets/src/views/index.vue

```diff
@@ -1,267 +1,80 @@
 <template>
 
-  <div class="setup">
-    <div class="wrapper" role="main">
-      <img class="logo" src="./icons/logo-min.png" alt="Kolibri logo">
+  <div class="onboarding">
+    <progress-toolbar
+      @backButtonClicked="goToPreviousStep"
+      :currentStep="onboardingStep"
+      :totalSteps="totalOnboardingSteps"/>
+
+    <component
+      :is="currentOnboardingForm"
+      :submitText="submitText"
+      :onboardingData="onboardingData"
```

Comment:
In general we should avoid passing objects as props because it obscures the true interface of a component. Always prefer simple prop types (bool, string, number, etc).

In this case, `onboardingData` seems like a thin wrapper around vuex state. In this case it would probably be cleaner to just reference the vuex state directly in each of the top-level "pages", because then the vuex getters are explicitly listed whereas here the attributes you need in each component are implicit.

Reply:
this is how I had it initially, actually!

I've been experimenting with [some advice in the flux documentation](https://facebook.github.io/flux/docs/in-depth-overview.html#views-and-controller-views) :
> We often pass the entire state of the store down the chain of views in a single object, allowing different descendants to use what they need. In addition to keeping the controller-like behavior at the top of the hierarchy, and thus keeping our descendant views as functionally pure as possible, passing down the entire state of the store in a single object also has the effect of reducing the number of props we need to manage.

You have any thoughts on their reasoning?

Reply:
Their advice is generally sound, but I don't necessarily agree with all of it.

There are at least three (sometimes conflicting!) priorities to optimize for when designing a component:

* Composability
  * Related to how decoupled a component is from external state - how "pure" it is. This affects how easily other parts of the app can be changed without having side effects on this component.
  * Also related to a having a well-defined interface.
  * A pure component with a well-defined interface can be easily re-used in other parts of the app with different data.
* Well-defined dependencies
  * Especially with dynamically typed languages, being explicit about dependencies is critical to allow future refactoring. If you can't easily search for where some state is used, there's no way to know if changing that logic will cause regressions.
  * A pure component with no external dependencies is usually - but not always - the best strategy.
  * If a pure component is not possible, external dependencies should be enumerated as explicitly as possible
* Minimizing boilerplate
  * Code is redundant or repetitive can be difficult to refactor
  * Boilerplate leads to copy-paste coding which can introduce bugs
  * When a large percentage of code is boilerplate and doesn't relate to the core logic of a component, it hinders readability

Different situations lend to different relative priorities. For example, the top-level "page" component in an app, a widely shared core widget, and a single-use helper component each has different forces affecting the design. 

-----

So with regard to their specific advice:

> Close to the top of the nested view hierarchy, a special kind of view listens for events that are broadcast by the stores that it depends on. We call this a controller-view, as it provides the glue code to get the data from the stores and to pass this data down the chain of its descendants. We might have one of these controller-views governing any significant section of the page.

This is an excellent strategy that have been actively trying to follow in Kolibri: try to keep Vuex getters high in the component hierarchy, and pass props down to keep sub-components composable and decoupled.

In this particular case, the individual pages (`defaultLanguageForm`, `facilityNameForm`, etc) are more appropriate as the "controller-view" glue components:

* each of them requires a specific subset of all possible page states. By moving the vuex getters to the individual pages, the dependencies become much better-defined
* passing the entire page into each of the top level pages doesn't do anything useful in terms of composability or well-defined interfaces, and it adds boilerplate.

> We often pass the entire state of the store down the chain of views in a single object, allowing different descendants to use what they need. In addition to keeping the controller-like behavior at the top of the hierarchy, and thus keeping our descendant views as functionally pure as possible, passing down the entire state of the store in a single object also has the effect of reducing the number of props we need to manage.

I don't agree with this strategy. They seem to be optimizing primarily for reduced boilerplate at the cost of well-defined interfaces and explicit dependencies. Components designed like this might be technically "pure", but passing the complete page state object down the component hierarchy does nothing to help with composability or well-defined dependencies.

Note that any refactoring at the app level might break the component - it's tightly coupled to the global application state schema. Furthermore the particular sub-state that the component is dependent on is defined implicitly rather than enumerated explicitly, so it's hard to even know if the component would break due to some particular app-level refactor.

Reply:
Thanks a lot for the in-depth analysis. Will try and keep all this stuff in mind and develop my own opinions as time goes on.